File: tools/codebox/runtime/runners/py/telemetry.py
#!/usr/bin/env python3
"""codebox-run-py - OpenTelemetry wrapper for Python.

The real code is never imported: it runs only via subprocess spawn, as a
child process, while this wrapper records spans + metrics with the
OpenTelemetry Python SDK and writes a ``codebox.telemetry/v1`` JSON document
to $CODEBOX_TELEMETRY_OUT.

Usage: codebox-run-py <file.py> [args...]
"""
from __future__ import annotations

import json
import os
import resource
import signal
import socket
import subprocess
import sys
import tempfile
import time
from datetime import datetime, timezone
from importlib.metadata import version

from opentelemetry import trace
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import InMemoryMetricReader
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import SimpleSpanProcessor
from opentelemetry.sdk.trace.export.in_memory_span_exporter import InMemorySpanExporter
from opentelemetry.trace import SpanKind, Status, StatusCode

RUNNER = {
    "language": "python",
    "name": "codebox-run-py",
    "sdk": "opentelemetry-python",
    "sdk_version": version("opentelemetry-sdk"),
}


def iso(ts: float) -> str:
    return datetime.fromtimestamp(ts, tz=timezone.utc).isoformat()


def serialize_span(span) -> dict:
    ctx = span.get_span_context()
    return {
        "name": span.name,
        "trace_id": format(ctx.trace_id, "032x"),
        "span_id": format(ctx.span_id, "016x"),
        "parent_span_id": format(span.parent.span_id, "016x") if span.parent else None,
        "kind": span.kind.name,
        "start_time_unix_nano": str(span.start_time),
        "end_time_unix_nano": str(span.end_time),
        "duration_ms": (span.end_time - span.start_time) / 1e6,
        "status": {"code": span.status.status_code.name, "message": span.status.description},
        "attributes": dict(span.attributes or {}),
        "events": [{"name": e.name, "attributes": dict(e.attributes or {})} for e in span.events],
    }


def serialize_metrics(reader: InMemoryMetricReader) -> list[dict]:
    data = reader.get_metrics_data()
    out: list[dict] = []
    if data is None:
        return out
    for rm in data.resource_metrics:
        for sm in rm.scope_metrics:
            for metric in sm.metrics:
                points = []
                for point in metric.data.data_points:
                    value = getattr(point, "value", None)
                    if value is None:  # histogram
                        value = {
                            "count": point.count,
                            "sum": point.sum,
                            "min": point.min,
                            "max": point.max,
                            "buckets": {"boundaries": list(point.explicit_bounds), "counts": list(point.bucket_counts)},
                        }
                    points.append({"attributes": dict(point.attributes or {}), "value": value})
                out.append({
                    "name": metric.name,
                    "unit": metric.unit,
                    "type": type(metric.data).__name__.upper(),
                    "data_points": points,
                })
    return out


def main() -> int:
    if len(sys.argv) < 2:
        print("usage: codebox-run-py <file.py> [args...]", file=sys.stderr)
        return 64

    entry = os.path.abspath(sys.argv[1])
    args = sys.argv[2:]
    out_file = os.environ.get("CODEBOX_TELEMETRY_OUT") or os.path.join(
        tempfile.gettempdir(), f"codebox-telemetry-{os.getpid()}.json"
    )
    timeout_ms = int(os.environ.get("CODEBOX_TIMEOUT_MS") or 0)
    python = os.environ.get("CODEBOX_PYTHON") or sys.executable
    command = [python, entry, *args]

    otel_resource = Resource.create({
        "service.name": os.environ.get("OTEL_SERVICE_NAME") or f"codebox-{os.environ.get('CODEBOX_BOX', 'local')}",
        "service.version": "1.0.0",
        "codebox.box": os.environ.get("CODEBOX_BOX", ""),
        "codebox.run_id": os.environ.get("CODEBOX_RUN_ID", ""),
        "process.runtime.name": sys.implementation.name,
        "process.runtime.version": sys.version.split()[0],
        "host.name": socket.gethostname(),
    })

    exporter = InMemorySpanExporter()
    tracer_provider = TracerProvider(resource=otel_resource)
    tracer_provider.add_span_processor(SimpleSpanProcessor(exporter))
    tracer = tracer_provider.get_tracer(RUNNER["name"], "1.0.0")

    reader = InMemoryMetricReader()
    meter_provider = MeterProvider(resource=otel_resource, metric_readers=[reader])
    meter = meter_provider.get_meter(RUNNER["name"], "1.0.0")
    duration_hist = meter.create_histogram("process.duration", unit="ms", description="Wall time of the spawned process")
    cpu_counter = meter.create_counter("process.cpu.time", unit="ms", description="CPU time of the spawned process")
    rss_gauge = meter.create_gauge("process.memory.max_rss", unit="KiBy", description="Peak resident set size")
    faults_counter = meter.create_counter("process.paging.faults", unit="{fault}")
    ctx_counter = meter.create_counter("process.context_switches", unit="{count}")

    attrs = {"code.filepath": entry}
    exit_code: int | None = None
    signal_name: str | None = None
    timed_out = False
    spawn_error: str | None = None
    pid: int | None = None

    before = resource.getrusage(resource.RUSAGE_CHILDREN)
    started = time.time()
    t0 = time.perf_counter_ns()

    with tracer.start_as_current_span("codebox.run", kind=SpanKind.INTERNAL,
                                      attributes={**attrs, "codebox.language": "python"}) as root:
        with tracer.start_as_current_span("process.exec", kind=SpanKind.CLIENT,
                                          attributes={"process.command": command[0],
                                                      "process.command_args": command}) as exec_span:
            try:
                child = subprocess.Popen(command)
                pid = child.pid
                exec_span.set_attribute("process.pid", pid)
                try:
                    returncode = child.wait(timeout=timeout_ms / 1000 if timeout_ms > 0 else None)
                except subprocess.TimeoutExpired:
                    timed_out = True
                    child.kill()
                    returncode = child.wait()
                if returncode < 0:
                    signal_name = signal.Signals(-returncode).name
                else:
                    exit_code = returncode
            except OSError as err:
                spawn_error = str(err)

            exec_span.set_attribute("process.exit.code", exit_code if exit_code is not None else -1)
            if signal_name:
                exec_span.set_attribute("process.exit.signal", signal_name)
            if exit_code == 0:
                exec_span.set_status(Status(StatusCode.OK))
            else:
                exec_span.set_status(Status(StatusCode.ERROR,
                                            spawn_error or ("timeout" if timed_out else f"exit {exit_code or signal_name}")))
        root.set_attribute("process.exit.code", exit_code if exit_code is not None else -1)
        root.set_status(Status(StatusCode.OK) if exit_code == 0 else Status(StatusCode.ERROR))

    duration_ms = (time.perf_counter_ns() - t0) / 1e6
    ended = time.time()
    after = resource.getrusage(resource.RUSAGE_CHILDREN)

    usage = {
        "user_cpu_ms": (after.ru_utime - before.ru_utime) * 1000,
        "system_cpu_ms": (after.ru_stime - before.ru_stime) * 1000,
        "max_rss_kb": after.ru_maxrss,
        "minor_page_faults": after.ru_minflt - before.ru_minflt,
        "major_page_faults": after.ru_majflt - before.ru_majflt,
        "voluntary_ctx_switches": after.ru_nvcsw - before.ru_nvcsw,
        "involuntary_ctx_switches": after.ru_nivcsw - before.ru_nivcsw,
        "fs_in_blocks": after.ru_inblock - before.ru_inblock,
        "fs_out_blocks": after.ru_oublock - before.ru_oublock,
        "threads": None,
    }

    duration_hist.record(duration_ms, attrs)
    cpu_counter.add(usage["user_cpu_ms"], {**attrs, "cpu.mode": "user"})
    cpu_counter.add(usage["system_cpu_ms"], {**attrs, "cpu.mode": "system"})
    rss_gauge.set(usage["max_rss_kb"], attrs)
    faults_counter.add(usage["minor_page_faults"], {**attrs, "process.paging.fault_type": "minor"})
    faults_counter.add(usage["major_page_faults"], {**attrs, "process.paging.fault_type": "major"})
    ctx_counter.add(usage["voluntary_ctx_switches"], {**attrs, "process.context_switch_type": "voluntary"})
    ctx_counter.add(usage["involuntary_ctx_switches"], {**attrs, "process.context_switch_type": "involuntary"})

    document = {
        "schema": "codebox.telemetry/v1",
        "runner": RUNNER,
        "run_id": os.environ.get("CODEBOX_RUN_ID"),
        "box": os.environ.get("CODEBOX_BOX"),
        "entry": entry,
        "command": command,
        "compile": None,
        "process": {
            "pid": pid,
            "exit_code": exit_code,
            "signal": signal_name,
            "timed_out": timed_out,
            "spawn_error": spawn_error,
            "started_at": iso(started),
            "ended_at": iso(ended),
            "duration_ms": duration_ms,
        },
        "resources": usage,
        "otel": {
            "resource": dict(otel_resource.attributes),
            "spans": [serialize_span(s) for s in exporter.get_finished_spans()],
            "metrics": serialize_metrics(reader),
        },
    }

    os.makedirs(os.path.dirname(os.path.abspath(out_file)), exist_ok=True)
    with open(out_file, "w", encoding="utf-8") as fh:
        json.dump(document, fh, indent=2, default=str)
    tracer_provider.shutdown()
    meter_provider.shutdown()

    if exit_code is not None:
        return exit_code
    return 128 + signal.Signals[signal_name].value if signal_name else 1


if __name__ == "__main__":
    sys.exit(main())

File: tools/codebox/runtime/runners/ts/telemetry.ts
/**
 * codebox-run-ts — OpenTelemetry wrapper for TypeScript / JavaScript.
 *
 * The real code is never imported: it only runs via spawn(), as a child
 * process, while this wrapper records spans + metrics with the OpenTelemetry
 * JS SDK (v2) and writes a `codebox.telemetry/v1` JSON document to
 * $CODEBOX_TELEMETRY_OUT.
 *
 * Usage: codebox-run-ts <file.ts|file.js> [args...]
 */
import { spawn } from 'node:child_process';
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { SpanKind, SpanStatusCode, context, trace } from '@opentelemetry/api';
import { resourceFromAttributes } from '@opentelemetry/resources';
import {
  AggregationTemporality,
  DataPointType,
  MeterProvider,
  MetricReader,
} from '@opentelemetry/sdk-metrics';
import {
  BasicTracerProvider,
  InMemorySpanExporter,
  SimpleSpanProcessor,
  type ReadableSpan,
} from '@opentelemetry/sdk-trace-base';
import { ATTR_SERVICE_NAME, ATTR_SERVICE_VERSION } from '@opentelemetry/semantic-conventions';

const RUNNER = { language: 'typescript', name: 'codebox-run-ts', sdk: 'opentelemetry-js', sdk_version: '2.x' };

class ManualReader extends MetricReader {
  constructor() {
    super({ aggregationTemporalitySelector: () => AggregationTemporality.CUMULATIVE });
  }
  protected async onForceFlush(): Promise<void> {}
  protected async onShutdown(): Promise<void> {}
}

type ProcSample = { maxRssKb: number | null; volCtx: number | null; involCtx: number | null; threads: number | null };

const [entryArg, ...childArgs] = process.argv.slice(2);
if (!entryArg) {
  process.stderr.write('usage: codebox-run-ts <file> [args...]\n');
  process.exit(64);
}

const entry = path.resolve(entryArg);
const outFile = process.env.CODEBOX_TELEMETRY_OUT || path.join(os.tmpdir(), `codebox-telemetry-${process.pid}.json`);
const timeoutMs = Number(process.env.CODEBOX_TIMEOUT_MS || 0);

const resource = resourceFromAttributes({
  [ATTR_SERVICE_NAME]: process.env.OTEL_SERVICE_NAME || `codebox-${process.env.CODEBOX_BOX || 'local'}`,
  [ATTR_SERVICE_VERSION]: '1.0.0',
  'codebox.box': process.env.CODEBOX_BOX || '',
  'codebox.run_id': process.env.CODEBOX_RUN_ID || '',
  'process.runtime.name': 'nodejs',
  'process.runtime.version': process.versions.node,
  'host.name': os.hostname(),
});

const spanExporter = new InMemorySpanExporter();
const tracerProvider = new BasicTracerProvider({ resource, spanProcessors: [new SimpleSpanProcessor(spanExporter)] });
const tracer = tracerProvider.getTracer(RUNNER.name, '1.0.0');

const metricReader = new ManualReader();
const meterProvider = new MeterProvider({ resource, readers: [metricReader] });
const meter = meterProvider.getMeter(RUNNER.name, '1.0.0');
const durationHistogram = meter.createHistogram('process.duration', { unit: 'ms', description: 'Wall time of the spawned process' });
const cpuCounter = meter.createCounter('process.cpu.time', { unit: 'ms', description: 'CPU time of the spawned process' });
const rssGauge = meter.createGauge('process.memory.max_rss', { unit: 'KiBy', description: 'Peak resident set size' });
const faultsCounter = meter.createCounter('process.paging.faults', { unit: '{fault}' });

/** Children CPU/fault counters of *this* process (fields of /proc/self/stat), in clock ticks. */
function selfChildrenStat(): { cutime: number; cstime: number; cminflt: number; cmajflt: number } | null {
  try {
    const raw = fs.readFileSync('/proc/self/stat', 'utf8');
    const fields = raw.slice(raw.lastIndexOf(')') + 2).split(' ');
    // fields[0] is field 3 (state); field N => fields[N - 3]
    return {
      cminflt: Number(fields[8]),
      cmajflt: Number(fields[10]),
      cutime: Number(fields[13]),
      cstime: Number(fields[14]),
    };
  } catch {
    return null;
  }
}

function sampleChild(pid: number, current: ProcSample): ProcSample {
  try {
    const status = fs.readFileSync(`/proc/${pid}/status`, 'utf8');
    const pick = (key: string) => {
      const match = new RegExp(`^${key}:\\s+(\\d+)`, 'm').exec(status);
      return match ? Number(match[1]) : null;
    };
    const hwm = pick('VmHWM');
    return {
      maxRssKb: hwm !== null ? Math.max(hwm, current.maxRssKb ?? 0) : current.maxRssKb,
      volCtx: pick('voluntary_ctxt_switches') ?? current.volCtx,
      involCtx: pick('nonvoluntary_ctxt_switches') ?? current.involCtx,
      threads: pick('Threads') ?? current.threads,
    };
  } catch {
    return current;
  }
}

function commandFor(file: string, args: string[]): string[] {
  if (/\.(ts|mts|cts|tsx)$/.test(file)) {
    const tsxLoader = import.meta.resolve('tsx');
    return [process.execPath, '--import', tsxLoader, file, ...args];
  }
  return [process.execPath, file, ...args];
}

function hrToNanos([seconds, nanos]: [number, number]): string {
  return (BigInt(seconds) * 1_000_000_000n + BigInt(nanos)).toString();
}

function serializeSpan(span: ReadableSpan) {
  const ctx = span.spanContext();
  return {
    name: span.name,
    trace_id: ctx.traceId,
    span_id: ctx.spanId,
    parent_span_id: span.parentSpanContext?.spanId ?? null,
    kind: SpanKind[span.kind],
    start_time_unix_nano: hrToNanos(span.startTime),
    end_time_unix_nano: hrToNanos(span.endTime),
    duration_ms: span.duration[0] * 1e3 + span.duration[1] / 1e6,
    status: { code: SpanStatusCode[span.status.code], message: span.status.message ?? null },
    attributes: span.attributes,
    events: span.events.map((e) => ({ name: e.name, attributes: e.attributes ?? {} })),
  };
}

async function collectMetrics() {
  const { resourceMetrics } = await metricReader.collect();
  return resourceMetrics.scopeMetrics.flatMap((scope) =>
    scope.metrics.map((metric) => ({
      name: metric.descriptor.name,
      unit: metric.descriptor.unit,
      type: DataPointType[metric.dataPointType],
      data_points: metric.dataPoints.map((point) => ({ attributes: point.attributes, value: point.value })),
    })),
  );
}

async function main() {
  const command = commandFor(entry, childArgs);
  const root = tracer.startSpan('codebox.run', {
    kind: SpanKind.INTERNAL,
    attributes: { 'code.filepath': entry, 'codebox.language': RUNNER.language },
  });

  const before = selfChildrenStat();
  const ticks = 100; // USER_HZ on Linux
  const startedAt = new Date();
  const t0 = process.hrtime.bigint();

  const rootContext = trace.setSpan(context.active(), root);
  const result = await new Promise<{ code: number | null; signal: NodeJS.Signals | null; timedOut: boolean; pid: number | null; sample: ProcSample; error?: string }>((resolve) => {
      const execSpan = tracer.startSpan(
        'process.exec',
        { kind: SpanKind.CLIENT, attributes: { 'process.command': command[0], 'process.command_args': command } },
        rootContext,
      );
      const child = spawn(command[0], command.slice(1), { stdio: 'inherit', env: process.env });
      let sample: ProcSample = { maxRssKb: null, volCtx: null, involCtx: null, threads: null };
      let timedOut = false;
      const sampler = setInterval(() => {
        if (child.pid) sample = sampleChild(child.pid, sample);
      }, 10);
      const killer = timeoutMs > 0
        ? setTimeout(() => {
            timedOut = true;
            child.kill('SIGKILL');
          }, timeoutMs)
        : null;
      if (child.pid) sample = sampleChild(child.pid, sample);
      execSpan.setAttribute('process.pid', child.pid ?? -1);

      const finish = (code: number | null, signal: NodeJS.Signals | null, error?: string) => {
        clearInterval(sampler);
        if (killer) clearTimeout(killer);
        execSpan.setAttribute('process.exit.code', code ?? -1);
        if (signal) execSpan.setAttribute('process.exit.signal', signal);
        if (error || code !== 0) {
          execSpan.setStatus({ code: SpanStatusCode.ERROR, message: error || (timedOut ? 'timeout' : `exit ${code ?? signal}`) });
        } else {
          execSpan.setStatus({ code: SpanStatusCode.OK });
        }
        execSpan.end();
        resolve({ code, signal, timedOut, pid: child.pid ?? null, sample, error });
      };
      child.on('error', (err) => finish(null, null, err.message));
      child.on('exit', (code, signal) => finish(code, signal));
    });

  const t1 = process.hrtime.bigint();
  const endedAt = new Date();
  const after = selfChildrenStat();
  const durationMs = Number(t1 - t0) / 1e6;
  const delta = (key: 'cutime' | 'cstime' | 'cminflt' | 'cmajflt') =>
    before && after ? after[key] - before[key] : null;
  const userMs = delta('cutime') !== null ? (delta('cutime') as number) * (1000 / ticks) : null;
  const sysMs = delta('cstime') !== null ? (delta('cstime') as number) * (1000 / ticks) : null;

  const attrs = { 'code.filepath': entry };
  durationHistogram.record(durationMs, attrs);
  if (userMs !== null) cpuCounter.add(userMs, { ...attrs, 'cpu.mode': 'user' });
  if (sysMs !== null) cpuCounter.add(sysMs, { ...attrs, 'cpu.mode': 'system' });
  if (result.sample.maxRssKb !== null) rssGauge.record(result.sample.maxRssKb, attrs);
  if (delta('cminflt') !== null) faultsCounter.add(delta('cminflt') as number, { ...attrs, 'process.paging.fault_type': 'minor' });
  if (delta('cmajflt') !== null) faultsCounter.add(delta('cmajflt') as number, { ...attrs, 'process.paging.fault_type': 'major' });

  root.setAttribute('process.exit.code', result.code ?? -1);
  root.setStatus(result.code === 0 ? { code: SpanStatusCode.OK } : { code: SpanStatusCode.ERROR, message: result.error || `exit ${result.code ?? result.signal}` });
  root.end();

  const document = {
    schema: 'codebox.telemetry/v1',
    runner: RUNNER,
    run_id: process.env.CODEBOX_RUN_ID || null,
    box: process.env.CODEBOX_BOX || null,
    entry,
    command,
    compile: null,
    process: {
      pid: result.pid,
      exit_code: result.code,
      signal: result.signal,
      timed_out: result.timedOut,
      spawn_error: result.error ?? null,
      started_at: startedAt.toISOString(),
      ended_at: endedAt.toISOString(),
      duration_ms: durationMs,
    },
    resources: {
      user_cpu_ms: userMs,
      system_cpu_ms: sysMs,
      max_rss_kb: result.sample.maxRssKb,
      minor_page_faults: delta('cminflt'),
      major_page_faults: delta('cmajflt'),
      voluntary_ctx_switches: result.sample.volCtx,
      involuntary_ctx_switches: result.sample.involCtx,
      fs_in_blocks: null,
      fs_out_blocks: null,
      threads: result.sample.threads,
    },
    otel: {
      resource: resource.attributes,
      spans: spanExporter.getFinishedSpans().map(serializeSpan),
      metrics: await collectMetrics(),
    },
  };

  fs.mkdirSync(path.dirname(outFile), { recursive: true });
  fs.writeFileSync(outFile, JSON.stringify(document, null, 2));
  await Promise.all([tracerProvider.shutdown(), meterProvider.shutdown()]);
  process.exitCode = result.code ?? (result.signal ? 128 : 1);
}

main().catch((err) => {
  process.stderr.write(`codebox-run-ts: ${err instanceof Error ? err.stack : String(err)}\n`);
  process.exit(70);
});
